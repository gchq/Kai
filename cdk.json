{
  "app": "npx ts-node bin/kai.ts",
  "context": {
    "@aws-cdk/core:enableStackNameDuplicates": "true",
    "aws-cdk:enableDiffNoFail": "true",
    "vpcId": "DEFAULT",
    "extraIngressSecurityGroups": "",
    "globalTags": {}
  }
}
