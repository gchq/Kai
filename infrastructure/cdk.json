{
  "app": "npx ts-node bin/kai.ts",
  "context": {
    "vpcId": "DEFAULT",
    "extraIngressSecurityGroups": "",
    "globalTags": {},
    "databaseProps": {
      "graphTableScalingProps": {
        "minCapacity": 1,
        "maxCapacity": 25,
        "targetUtilizationPercent": 80
      },
      "namespaceTableScalingProps": {
        "minCapacity": 1,
        "maxCapacity": 25,
        "targetUtilizationPercent": 80
      }
    }
  }
}
