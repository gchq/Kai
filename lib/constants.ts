export const LAMBDA_LAYER_ARN = 'arn:aws:serverlessrepo:us-east-1:903779448426:applications/lambda-layer-kubectl'; 
export const LAMBDA_LAYER_VERSION = '2.0.0-beta3';