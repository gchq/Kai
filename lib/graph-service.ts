import core = require('@aws-cdk/core');
import lambda = require('@aws-cdk/aws-lambda');
import apigateway = require('@aws-cdk/aws-apigateway');

export class GraphService extends core.Construct {
    constructor(scope: core.Construct, id: string) {
        super(scope, id);

        const lambdas = new lambda.AssetCode('lambdas');

        const getAllGraphsHandler = new lambda.Function(this, 'GetAllGraphsHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.getAll'
        });

        const addGraphHandler = new lambda.Function(this, 'AddGraphHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.add'
        });

        const deleteGraphHandler = new lambda.Function(this, 'DeleteGraphHandler', {
            runtime: lambda.Runtime.NODEJS_10_X,
            code: lambdas,
            handler: 'graphs.delete'
        });

        const api = new apigateway.RestApi(this, 'graph-api', {
            restApiName: 'Graph Service',
            description: 'This service allows you to control graphs'
        });

        const getAllGraphsIntegration = new apigateway.LambdaIntegration(getAllGraphsHandler, {});
        const addGraphIntegration = new apigateway.LambdaIntegration(addGraphHandler, {});
        const deleteGraphIntegration = new apigateway.LambdaIntegration(deleteGraphHandler, {});

        api.root.addMethod('GET', getAllGraphsIntegration);
        api.root.addMethod('POST', addGraphIntegration);
        const graphEndpoint = api.root.addResource("{graphUrl}");
        graphEndpoint.addMethod('DELETE', deleteGraphIntegration);
    }
}