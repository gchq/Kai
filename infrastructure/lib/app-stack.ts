/*
 * Copyright 2020 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as cdk from "@aws-cdk/core";
import * as sam from "@aws-cdk/aws-sam";
import { GraphPlatForm } from "./platform/graph-platform";
import { GraphUninstaller } from "./platform/graph-uninstaller";
import { KaiRestApi } from "./rest-api/kai-rest-api";
import { LAMBDA_LAYER_ARN, LAMBDA_LAYER_VERSION, ADD_GRAPH_TIMEOUT, DELETE_GRAPH_TIMEOUT, DELETE_GRAPH_WORKER_BATCH_SIZE, ADD_GRAPH_WORKER_BATCH_SIZE } from "./constants";
import { LayerVersion } from "@aws-cdk/aws-lambda";
import { GraphDatabase } from "./database/graph-database";
import { Worker } from "./workers/worker";
import { KaiUserPool } from "./authentication/user-pool";
import { GraphDatabaseProps } from "./database/graph-database-props";
import { PolicyStatement } from "@aws-cdk/aws-iam";

// The main stack for Kai
export class AppStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: cdk.StackProps) {
        super(scope, id, props);

        // User Pool
        const userPool = new KaiUserPool(this, "KaiUserPool");

        // Graph Platform
        const platform = new GraphPlatForm(this, "GraphPlatform");

        // Graph Table
        const graphDBProps: GraphDatabaseProps = this.node.tryGetContext("graphDatabaseProps");
        const database = new GraphDatabase(this, "GraphDatabase", graphDBProps);

        // REST API
        const kaiRest = new KaiRestApi(this, "KaiRestApi", {
            graphTable: database.table,
            userPoolArn: userPool.userPoolArn,
            userPoolId: userPool.userPoolId
        });

        // Kubectl Lambda layer
        const samApp = new sam.CfnApplication(this, "SamLayer", {
            location: {
                applicationId: LAMBDA_LAYER_ARN,
                semanticVersion: LAMBDA_LAYER_VERSION
            },
            parameters: {
                LayerName: `${this.node.uniqueId}-kubectl-layer`
            }
        });

        const layerVersionArn = samApp.getAtt("Outputs.LayerVersionArn").toString();
        const kubectlLambdaLayer = LayerVersion.fromLayerVersionArn(this, "KubectlLambdaLayer", layerVersionArn);

        // Describe EKS cluster policy statement
        const describeClusterPolicyStatement = new PolicyStatement({
            actions: [ "eks:DescribeCluster" ],
            resources: [ platform.eksCluster.clusterArn ]
        });

        // Manage EBS Volumes policy statement
        const manageVolumesPolicyStatement = new PolicyStatement({
            resources: ["*"],
            actions: [
                "ec2:DescribeVolumes",
                "ec2:DeleteVolume"
            ]
        });

        // Workers
        new Worker(this, "AddGraphWorker", {
            cluster: platform.eksCluster,
            queue: kaiRest.addGraphQueue,
            kubectlLayer: kubectlLambdaLayer,
            graphTable: database.table,
            handler: "add_graph.handler",
            timeout: ADD_GRAPH_TIMEOUT,
            batchSize: ADD_GRAPH_WORKER_BATCH_SIZE,
            policyStatements: [
                describeClusterPolicyStatement
            ]
        });

        const deleteGraphWorker = new Worker(this, "DeleteGraphWorker", {
            cluster: platform.eksCluster,
            queue: kaiRest.deleteGraphQueue,
            kubectlLayer: kubectlLambdaLayer,
            graphTable: database.table,
            handler: "delete_graph.handler",
            timeout: DELETE_GRAPH_TIMEOUT,
            batchSize: DELETE_GRAPH_WORKER_BATCH_SIZE,
            policyStatements: [
                describeClusterPolicyStatement,
                manageVolumesPolicyStatement
            ]
        });

        // Graph uninstaller
        new GraphUninstaller(this, "GraphUninstaller", {
            getGraphsFunctionArn: kaiRest.getGraphsLambda.functionArn,
            deleteGraphFunctionArn: kaiRest.deleteGraphLambda.functionArn,
            kubectlLayer: kubectlLambdaLayer,
            timeout: cdk.Duration.seconds(30),
            dependencies: [
                platform,
                database,
                deleteGraphWorker,
                kaiRest.getGraphsLambda,
                kaiRest.deleteGraphLambda,
                kaiRest.deleteGraphQueue
            ]
        });

        // Graph uninstaller
        new GraphUninstaller(this, "GraphUninstaller", {
            getGraphsFunctionArn: kaiRest.getGraphsLambda.functionArn,
            deleteGraphFunctionArn: kaiRest.deleteGraphLambda.functionArn,
            kubectlLayer: kubectlLambdaLayer,
            timeout: cdk.Duration.seconds(30),
            dependencies: [
                platform,
                database,
                deleteGraphWorker,
                kaiRest.getGraphsLambda,
                kaiRest.deleteGraphLambda,
                kaiRest.deleteGraphQueue
            ]
        });
    }
}