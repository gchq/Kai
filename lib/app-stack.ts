import * as cdk from "@aws-cdk/core";
import * as sam from "@aws-cdk/aws-sam";
import { GraphPlatForm } from "./platform/graph-platform";
import { KaiRestApi } from "./rest-api/kai-rest-api";
import { LAMBDA_LAYER_ARN, LAMBDA_LAYER_VERSION } from "./constants";
import { LayerVersion } from "@aws-cdk/aws-lambda";
import { AddGraphWorker } from "./workers/add-graph-worker";

// The main stack for Kai
export class AppStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: cdk.StackProps) {
        super(scope, id, props);

        // Graph Platform
        const platform = new GraphPlatForm(this, "GraphPlatform");

        // REST API
        const kaiRest = new KaiRestApi(this, "KaiRestApi");

        // Kubectl Lambda layer
        const samApp = new sam.CfnApplication(this, "SamLayer", {
            location: {
            applicationId: LAMBDA_LAYER_ARN,
            semanticVersion: LAMBDA_LAYER_VERSION
            },
            parameters: {
            LayerName: `${this.node.uniqueId}-kubectl-layer`
            }
        });

        const layerVersionArn = samApp.getAtt("Outputs.LayerVersionArn").toString();
        const kubectlLambdaLayer = LayerVersion.fromLayerVersionArn(this, "KubectlLambdaLayer", layerVersionArn);

        // Workers
        new AddGraphWorker(this, "AddGraphWorker", {
            cluster: platform.eksCluster,
            queue: kaiRest.addGraphQueue,
            kubectlLayer: kubectlLambdaLayer
        });
    }
}