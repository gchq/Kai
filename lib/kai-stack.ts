import cdk = require('@aws-cdk/core');
import { GraphService } from './graph-service';

export class KaiStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    new GraphService(this, 'Graphs');


  }
}
