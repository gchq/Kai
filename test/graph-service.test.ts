import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import cdk = require('@aws-cdk/core');
import { GraphService } from '../lib/graph-service';

test('GraphService matches the snapshot', () => {
    // Given
    const stack = new cdk.Stack();
    // When
    new GraphService(stack, "GraphService");
    // Then
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('GraphService should have a GetAll endpoint', () => {
    // Given
    const stack = new cdk.Stack();
    // When
    new GraphService(stack, "GraphService");
    // Then
    expect(stack).toHaveResource("AWS::Lambda::Function", {
        Handler: "graphs.getAll"
    });
});

test('GraphService should have a Delete endpoint', () => {
    // Given
    const stack = new cdk.Stack();
    // When
    new GraphService(stack, "GraphService");
    // Then
    expect(stack).toHaveResource("AWS::Lambda::Function", {
        Handler: "graphs.delete"
    });
});

test('GraphService should have an Add endpoint', () => {
    // Given
    const stack = new cdk.Stack();
    // When
    new GraphService(stack, "GraphService");
    // Then
    expect(stack).toHaveResource("AWS::Lambda::Function", {
        Handler: "graphs.add"
    });
});
