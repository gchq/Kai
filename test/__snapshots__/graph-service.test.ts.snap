// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphService matches the snapshot 1`] = `
Object {
  "Outputs": Object {
    "GraphServicegraphapiEndpoint71D8AA38": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "GraphServicegraphapiB91541B3",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "GraphServicegraphapiDeploymentStageprod5D5652C9",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957ArtifactHashFE671783": Object {
      "Description": "Artifact hash for asset \\"0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957\\"",
      "Type": "String",
    },
    "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3Bucket1888D2DC": Object {
      "Description": "S3 bucket for asset \\"0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957\\"",
      "Type": "String",
    },
    "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2": Object {
      "Description": "S3 key for asset version \\"0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GraphServiceAddGraphHandlerBAAE48B6": Object {
      "DependsOn": Array [
        "GraphServiceAddGraphHandlerServiceRole2E960413",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3Bucket1888D2DC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "graphs.add",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GraphServiceAddGraphHandlerServiceRole2E960413",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GraphServiceAddGraphHandlerServiceRole2E960413": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphServiceDeleteGraphHandlerC7A51AF9": Object {
      "DependsOn": Array [
        "GraphServiceDeleteGraphHandlerServiceRole4F4B2F61",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3Bucket1888D2DC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "graphs.delete",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GraphServiceDeleteGraphHandlerServiceRole4F4B2F61",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GraphServiceDeleteGraphHandlerServiceRole4F4B2F61": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphServiceGetAllGraphsHandler520E16D1": Object {
      "DependsOn": Array [
        "GraphServiceGetAllGraphsHandlerServiceRoleC21D818A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3Bucket1888D2DC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0186268f3ac97a795930ed7e9089409e8feb8a138043f95da96748560658d957S3VersionKey5BB03EF2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "graphs.getAll",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GraphServiceGetAllGraphsHandlerServiceRoleC21D818A",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "GraphServiceGetAllGraphsHandlerServiceRoleC21D818A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphServicegraphapiAccount943AB11A": Object {
      "DependsOn": Array [
        "GraphServicegraphapiB91541B3",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "GraphServicegraphapiCloudWatchRole06DF722A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "GraphServicegraphapiB91541B3": Object {
      "Properties": Object {
        "Description": "This service allows you to control graphs",
        "Name": "Graph Service",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "GraphServicegraphapiCloudWatchRole06DF722A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GraphServicegraphapiDeploymentEC5A93F2e298ae45995eae5b387d938113cdc74a": Object {
      "DependsOn": Array [
        "GraphServicegraphapigraphUrlDELETEB0D14381",
        "GraphServicegraphapigraphUrl1F95B55B",
        "GraphServicegraphapiGET9BB13121",
        "GraphServicegraphapiPOST339B5C53",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "GraphServicegraphapiDeploymentStageprod5D5652C9": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "GraphServicegraphapiDeploymentEC5A93F2e298ae45995eae5b387d938113cdc74a",
        },
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "GraphServicegraphapiGET9BB13121": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "GraphServiceGetAllGraphsHandler520E16D1",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "GraphServicegraphapiB91541B3",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "GraphServicegraphapiGETApiPermissionGraphServicegraphapi67232AF1GET64E97AA1": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceGetAllGraphsHandler520E16D1",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/",
              Object {
                "Ref": "GraphServicegraphapiDeploymentStageprod5D5652C9",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapiGETApiPermissionTestGraphServicegraphapi67232AF1GET8AE35598": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceGetAllGraphsHandler520E16D1",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapiPOST339B5C53": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "GraphServiceAddGraphHandlerBAAE48B6",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "GraphServicegraphapiB91541B3",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "GraphServicegraphapiPOSTApiPermissionGraphServicegraphapi67232AF1POST0D208D07": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceAddGraphHandlerBAAE48B6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/",
              Object {
                "Ref": "GraphServicegraphapiDeploymentStageprod5D5652C9",
              },
              "/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapiPOSTApiPermissionTestGraphServicegraphapi67232AF1POSTA74BFCB4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceAddGraphHandlerBAAE48B6",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/test-invoke-stage/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapigraphUrl1F95B55B": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "GraphServicegraphapiB91541B3",
            "RootResourceId",
          ],
        },
        "PathPart": "{graphUrl}",
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "GraphServicegraphapigraphUrlDELETEApiPermissionGraphServicegraphapi67232AF1DELETEgraphUrl6C68A167": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceDeleteGraphHandlerC7A51AF9",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/",
              Object {
                "Ref": "GraphServicegraphapiDeploymentStageprod5D5652C9",
              },
              "/DELETE/{graphUrl}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapigraphUrlDELETEApiPermissionTestGraphServicegraphapi67232AF1DELETEgraphUrlFFEBFC67": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GraphServiceDeleteGraphHandlerC7A51AF9",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "GraphServicegraphapiB91541B3",
              },
              "/test-invoke-stage/DELETE/{graphUrl}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "GraphServicegraphapigraphUrlDELETEB0D14381": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "DELETE",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "GraphServiceDeleteGraphHandlerC7A51AF9",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "GraphServicegraphapigraphUrl1F95B55B",
        },
        "RestApiId": Object {
          "Ref": "GraphServicegraphapiB91541B3",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
  },
}
`;
